File: services/room_manager.py
import time
import uuid
from typing import Dict, List, Optional

class RoomManager:
    def __init__(self):
        # In-memory storage for rooms
        # roomId -> { id, createdAt, participants: { id -> data }, messages: [], lastActivity }
        self.rooms: Dict[str, dict] = {}

    def create_room(self) -> dict:
        """Create a new room with a unique ID."""
        room_id = str(uuid.uuid4()).split('-')[0] + str(int(time.time()))[-5:]
        room_id = room_id[:13] # 13-character unique ID
        
        room = {
            "id": room_id,
            "createdAt": int(time.time() * 1000),
            "participants": {},
            "messages": [],
            "lastActivity": int(time.time() * 1000)
        }
        
        self.rooms[room_id] = room
        print(f"[RoomManager] Room created: {room_id}")
        
        return room

    def get_room(self, room_id: str) -> Optional[dict]:
        """Get room by ID."""
        return self.rooms.get(room_id)

    def add_participant(self, room_id: str, participant: dict) -> Optional[dict]:
        """Add participant to room."""
        room = self.get_room(room_id)
        if not room:
            return None

        participant_id = participant.get("id") or str(uuid.uuid4())
        participant_data = {
            **participant,
            "id": participant_id,
            "joinedAt": int(time.time() * 1000),
            "isOnline": True
        }
        
        room["participants"][participant_id] = participant_data
        room["lastActivity"] = int(time.time() * 1000)
        
        print(f"[RoomManager] Participant {participant_id} added to room {room_id}")
        
        return participant_data

    def remove_participant(self, room_id: str, participant_id: str) -> bool:
        """Remove participant from room."""
        room = self.get_room(room_id)
        if not room or not participant_id:
            return False

        if participant_id in room["participants"]:
            del room["participants"][participant_id]
            room["lastActivity"] = int(time.time() * 1000)
            print(f"[RoomManager] Participant {participant_id} removed from room {room_id}")
            return True
        
        return False

    def mark_participant_offline(self, room_id: str, participant_id: str) -> bool:
        """Mark participant as offline."""
        room = self.get_room(room_id)
        if not room:
            return False

        participant = room["participants"].get(participant_id)
        if participant:
            participant["isOnline"] = False
            print(f"[RoomManager] Participant {participant_id} marked offline in room {room_id}")
            return True
        
        return False

    def get_room_participants(self, room_id: str) -> List[dict]:
        """Get all participants in a room."""
        room = self.get_room(room_id)
        if not room:
            return []

        return list(room["participants"].values())

    def get_room_messages(self, room_id: str, limit: int = 500) -> List[dict]:
        """Get all messages in a room."""
        room = self.get_room(room_id)
        if not room:
            return []

        # Return most recent messages up to limit
        return room["messages"][-limit:]

    def add_message(self, room_id: str, message: dict) -> Optional[dict]:
        """Add message to room."""
        room = self.get_room(room_id)
        if not room:
            return None

        message_data = {
            **message,
            "id": message.get("id") or str(uuid.uuid4()),
            "timestamp": message.get("timestamp") or int(time.time() * 1000)
        }

        room["messages"].append(message_data)
        room["lastActivity"] = int(time.time() * 1000)

        # Limit message history to 500 messages per room
        if len(room["messages"]) > 500:
            room["messages"] = room["messages"][-500:]

        print(f"[RoomManager] Message added to room {room_id}")
        
        return message_data

    def get_all_rooms(self) -> List[dict]:
        """Get all active rooms."""
        return list(self.rooms.values())

    def delete_room(self, room_id: str) -> bool:
        """Delete room."""
        if room_id in self.rooms:
            del self.rooms[room_id]
            print(f"[RoomManager] Room {room_id} deleted")
            return True
        return False

    def cleanup_inactive_rooms(self) -> int:
        """Clean up inactive rooms (no participants for 24 hours)."""
        now = int(time.time() * 1000)
        inactive_threshold = 24 * 60 * 60 * 1000 # 24 hours
        cleaned_count = 0

        rooms_to_delete = []
        for room_id, room in self.rooms.items():
            has_no_participants = len(room["participants"]) == 0
            is_inactive = (now - room["lastActivity"]) > inactive_threshold

            if has_no_participants and is_inactive:
                rooms_to_delete.append(room_id)

        for room_id in rooms_to_delete:
            self.delete_room(room_id)
            cleaned_count += 1

        if cleaned_count > 0:
            print(f"[RoomManager] Cleaned up {cleaned_count} inactive rooms")

        return cleaned_count

# Singleton instance
room_manager = RoomManager()

File: services/RoomManager.js
const { v4: uuidv4 } = require('uuid');

/**
 * RoomManager Service
 * Handles room creation, participant tracking, and message storage
 */

class RoomManager {
    constructor() {
        // In-memory storage for rooms
        this.rooms = new Map();
    }

    /**
     * Create a new room with unique ID
     * @returns {Object} Room object with id, createdAt, participants, messages
     */
    createRoom() {
        const roomId = uuidv4().substring(0, 13); // 13-character unique ID
        const room = {
            id: roomId,
            createdAt: Date.now(),
            participants: new Map(),
            messages: [],
            lastActivity: Date.now()
        };
        
        this.rooms.set(roomId, room);
        console.log(`[RoomManager] Room created: ${roomId}`);
        
        return room;
    }

    /**
     * Get room by ID
     * @param {string} roomId - Room identifier
     * @returns {Object|null} Room object or null if not found
     */
    getRoom(roomId) {
        return this.rooms.get(roomId) || null;
    }

    /**
     * Add participant to room
     * @param {string} roomId - Room identifier
     * @param {Object} participant - Participant data
     * @returns {Object|null} Participant object or null if room not found
     */
    addParticipant(roomId, participant) {
        const room = this.getRoom(roomId);
        
        if (!room) {
            return null;
        }

        const participantId = participant.id || uuidv4();
        const participantData = {
            ...participant,
            id: participantId,
            joinedAt: Date.now(),
            isOnline: true
        };
        
        room.participants.set(participantId, participantData);
        room.lastActivity = Date.now();
        
        console.log(`[RoomManager] Participant ${participantId} added to room ${roomId}`);
        
        return participantData;
    }

    /**
     * Remove participant from room
     * @param {string} roomId - Room identifier
     * @param {string} participantId - Participant identifier
     * @returns {boolean} True if removed, false if not found
     */
    removeParticipant(roomId, participantId) {
        const room = this.getRoom(roomId);
        
        if (!room || !participantId) {
            return false;
        }

        const removed = room.participants.delete(participantId);
        
        if (removed) {
            room.lastActivity = Date.now();
            console.log(`[RoomManager] Participant ${participantId} removed from room ${roomId}`);
        }
        
        return removed;
    }

    /**
     * Mark participant as offline
     * @param {string} roomId - Room identifier
     * @param {string} participantId - Participant identifier
     * @returns {boolean} True if updated, false if not found
     */
    markParticipantOffline(roomId, participantId) {
        const room = this.getRoom(roomId);
        
        if (!room) {
            return false;
        }

        const participant = room.participants.get(participantId);
        
        if (participant) {
            participant.isOnline = false;
            console.log(`[RoomManager] Participant ${participantId} marked offline in room ${roomId}`);
            return true;
        }
        
        return false;
    }

    /**
     * Get all participants in a room
     * @param {string} roomId - Room identifier
     * @returns {Array} Array of participant objects
     */
    getRoomParticipants(roomId) {
        const room = this.getRoom(roomId);
        
        if (!room) {
            return [];
        }

        return Array.from(room.participants.values());
    }

    /**
     * Get all messages in a room
     * @param {string} roomId - Room identifier
     * @param {number} limit - Maximum number of messages to return (default: 500)
     * @returns {Array} Array of message objects
     */
    getRoomMessages(roomId, limit = 500) {
        const room = this.getRoom(roomId);
        
        if (!room) {
            return [];
        }

        // Return most recent messages up to limit
        return room.messages.slice(-limit);
    }

    /**
     * Add message to room
     * @param {string} roomId - Room identifier
     * @param {Object} message - Message object
     * @returns {Object|null} Message object with ID and timestamp, or null if room not found
     */
    addMessage(roomId, message) {
        const room = this.getRoom(roomId);
        
        if (!room) {
            return null;
        }

        const messageData = {
            ...message,
            id: message.id || uuidv4(),
            timestamp: message.timestamp || Date.now()
        };

        room.messages.push(messageData);
        room.lastActivity = Date.now();

        // Limit message history to 500 messages per room
        if (room.messages.length > 500) {
            room.messages = room.messages.slice(-500);
        }

        console.log(`[RoomManager] Message added to room ${roomId}`);
        
        return messageData;
    }

    /**
     * Get all active rooms
     * @returns {Array} Array of room objects
     */
    getAllRooms() {
        return Array.from(this.rooms.values());
    }

    /**
     * Delete room
     * @param {string} roomId - Room identifier
     * @returns {boolean} True if deleted, false if not found
     */
    deleteRoom(roomId) {
        const deleted = this.rooms.delete(roomId);
        
        if (deleted) {
            console.log(`[RoomManager] Room ${roomId} deleted`);
        }
        
        return deleted;
    }

    /**
     * Clean up inactive rooms (no participants for 24 hours)
     * Should be called periodically
     */
    cleanupInactiveRooms() {
        const now = Date.now();
        const inactiveThreshold = 24 * 60 * 60 * 1000; // 24 hours
        let cleanedCount = 0;

        for (const [roomId, room] of this.rooms.entries()) {
            const hasNoParticipants = room.participants.size === 0;
            const isInactive = (now - room.lastActivity) > inactiveThreshold;

            if (hasNoParticipants && isInactive) {
                this.deleteRoom(roomId);
                cleanedCount++;
            }
        }

        if (cleanedCount > 0) {
            console.log(`[RoomManager] Cleaned up ${cleanedCount} inactive rooms`);
        }

        return cleanedCount;
    }
}

// Export singleton instance
const roomManager = new RoomManager();

module.exports = roomManager;
